src/internal/storage/fileset/writer.go

```diff
@@ -102,80 +101,82 @@ func (w *Writer) checkPath(prevIdx, idx *index.Index) error {
 	return nil
 }
 
-// Copy copies a file to the file set writer.
-func (w *Writer) Copy(file File, datum string) error {
-	idx := file.Index()
-	copyIdx := &index.Index{
-		Path: idx.Path,
+// Delete creates a delete operation for a file.
+func (w *Writer) Delete(path, datum string) error {
+	idx := &index.Index{
+		Path: path,
 		File: &index.File{
 			Datum: datum,
 		},
 	}
-	if err := w.nextIdx(copyIdx); err != nil {
+	if err := w.checkIndex(w.deleteIdx, idx); err != nil {
 		return err
 	}
-	// Copy the file data refs.
-	for _, dataRef := range idx.File.DataRefs {
-		w.sizeBytes += dataRef.SizeBytes
-		if err := w.cw.Copy(dataRef); err != nil {
-			return err
-		}
-	}
-	return nil
+	w.deleteIdx = idx
+	return w.deletive.WriteIndex(idx)
 }
 
-func (w *Writer) callback(annotations []*chunk.Annotation) error {
-	for _, annotation := range annotations {
-		idx := annotation.Data.(*index.Index)
-		if w.lastIdx == nil {
-			w.lastIdx = idx
-		}
-		if idx.Path != w.lastIdx.Path || idx.File.Datum != w.lastIdx.File.Datum {
-			if err := w.additive.WriteIndex(w.lastIdx); err != nil {
-				return err
-			}
-			if w.indexFunc != nil {
-				if err := w.indexFunc(w.lastIdx); err != nil {
-					return err
-				}
-			}
-			w.lastIdx = idx
+// Copy copies a file to the file set writer.
+func (w *Writer) Copy(file File, datum string) error {
+	idx := file.Index()
+	size := index.SizeBytes(idx)
+	if size >= int64(w.batchThreshold) && chunk.StableDataRefs(idx.File.DataRefs) {
```

Comment:
the stable condition is mainly for backwards compatibility, right? (And I guess if the batch threshold changes) Thought I saw a comment about that somewhere but maybe I was imagining it

Reply:
right, existing files were not written with the batch threshold in mind so they might not be eligible for copy by reference.